src/cpu/core_dyn_x86.cpp

```diff
@@ -263,6 +263,31 @@ Bits CPU_Core_Dyn_X86_Run(void) {
 	};
 	auto_dh_fpu fpu_saver;
 
+	class auto_fpu_sync {
+	public:
+		auto_fpu_sync () {
+			FPU_SetTag(dyn_dh_fpu.state.tag);
+			FPU_SetCW(dyn_dh_fpu.state.cw);
+			fpu.sw = dyn_dh_fpu.state.sw;
+			TOP = FPU_GET_TOP();
+			for(Bitu i = 0;i < 8;i++){
+				fpu.p_regs[STV(i)].m1 = *((uint32_t*)(dyn_dh_fpu.state.st_reg[i]+0));
+				fpu.p_regs[STV(i)].m2 = *((uint32_t*)(dyn_dh_fpu.state.st_reg[i]+4));
+				fpu.p_regs[STV(i)].m3 = *((uint16_t*)(dyn_dh_fpu.state.st_reg[i]+8));
+			}
+		}
+		~auto_fpu_sync () {
+			FPU_SET_TOP(TOP);
+			dyn_dh_fpu.state.tag = FPU_GetTag();
+			dyn_dh_fpu.state.cw = fpu.cw;
+			dyn_dh_fpu.state.sw = fpu.sw;
+			for(Bitu i = 0;i < 8;i++){
+				*((uint32_t*)(dyn_dh_fpu.state.st_reg[i]+0)) = fpu.p_regs[STV(i)].m1;
+				*((uint32_t*)(dyn_dh_fpu.state.st_reg[i]+4)) = fpu.p_regs[STV(i)].m2;
+				*((uint16_t*)(dyn_dh_fpu.state.st_reg[i]+8)) = fpu.p_regs[STV(i)].m3;
```

Comment:
The `dyn_dh_fpu.state.st_reg`s are a `uint8_t`s, so their addresses can only be type-pun'd up other 1-byte types. 

Type-punning up to `uint16_t*` and `uint32_t*` is going to be undefined behavior.

(However, address type-punning down from a larger types (say, say a uint32_t) down to `uint16_t*` and byte-types is OK).

Reply:
But maybe some good news - I wasn't able to find anything using the dyn FPU's state regs; so maybe try commening out those entirely:

```
	struct {
		uint16_t cw;
		uint32_t sw;
		uint32_t tag;
		uint32_t ip;
		uint32_t cs;
		uint32_t ea;
		uint32_t ds;
		// uint8_t st_reg[8][10]; Currently unused
	} state;
```

And then knock out the loads and saves.

Reply:
Confirmed:

``` text
../../src/cpu/core_dyn_x86.cpp:274:29: runtime error: load of misaligned address 0x556c0e31d50e for type 'uint32_t', which requires 4 byte alignment
0x556c0e31d50e: note: pointer points here
 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00
             ^ 
../../src/cpu/core_dyn_x86.cpp:275:29: runtime error: load of misaligned address 0x556c0e31d512 for type 'uint32_t', which requires 4 byte alignment
0x556c0e31d512: note: pointer points here
 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00
              ^ 
../../src/cpu/core_dyn_x86.cpp:285:50: runtime error: store to misaligned address 0x556c0e31d50e for type 'uint32_t', which requires 4 byte alignment
0x556c0e31d50e: note: pointer points here
 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00
             ^ 
../../src/cpu/core_dyn_x86.cpp:286:50: runtime error: store to misaligned address 0x556c0e31d512 for type 'uint32_t', which requires 4 byte alignment
0x556c0e31d512: note: pointer points here
 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00
```

Reply:
You are right I need to fix that.

Reply:
But hopefully they're not needed at all? 

(atleast on my side, it compiles and runs OK with ` st_reg[8][10]` commented out along with the two loops.

Reply:
st_reg is not used directly in code but we need this space for the dynamically generated code.
The dyn core generates `fnsave` to store the physical FPU state in the `state` variable.  And this requires 94-108 bytes depending on cpu.  If you comment that out, the `fnsave` instruction will corrupt data.